pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -980,3 +995,109 @@ func getOrCreateIstioMetadata(cluster *cluster.Cluster) *structpb.Struct {
 	}
 	return cluster.Metadata.FilterMetadata[util.IstioMetadataKey]
 }
+
+func (configgen *ConfigGeneratorImpl) buildInboundHBONEClusters(cb *ClusterBuilder, proxy *model.Proxy, instances []*model.ServiceInstance) []*cluster.Cluster {
+	if !proxy.EnableHBONE() {
+		return nil
+	}
+	clusters := make([]*cluster.Cluster, 0)
+	for _, i := range instances {
+		p := i.Endpoint.EndpointPort
+		name := fmt.Sprintf("inbound-hbone|%d", p)
+		clusters = append(clusters, cb.buildInternalListenerCluster(name, name).build())
+	}
+	return clusters
+}
+
+func (cb *ClusterBuilder) buildInternalListenerCluster(clusterName string, listenerName string) *MutableCluster {
+	clusterType := cluster.Cluster_STATIC
+	llb := util.BuildInternalEndpoint(listenerName, nil)
+	port := &model.Port{}
+	localCluster := cb.buildDefaultCluster(clusterName, clusterType, llb,
+		model.TrafficDirectionInbound, port, nil, nil)
+	// no TLS
+	localCluster.cluster.TransportSocketMatches = nil
+	localCluster.cluster.TransportSocket = InternalUpstreamSocket
+	return localCluster
+}
+
+var HboneOrPlaintextSocket = []*cluster.Cluster_TransportSocketMatch{
+	{
+		Name:            "hbone",
+		Match:           hboneTransportSocketMatch,
+		TransportSocket: InternalUpstreamSocket,
+	},
+	defaultTransportSocketMatch(),
+}
+
+var InternalUpstreamSocket = &core.TransportSocket{
+	Name: "envoy.transport_sockets.internal_upstream",
+	ConfigType: &core.TransportSocket_TypedConfig{TypedConfig: protoconv.MessageToAny(&internalupstream.InternalUpstreamTransport{
+		PassthroughMetadata: []*internalupstream.InternalUpstreamTransport_MetadataValueSource{
+			{
+				Kind: &metadata.MetadataKind{Kind: &metadata.MetadataKind_Host_{}},
+				Name: "tunnel",
+			},
+			{
+				Kind: &metadata.MetadataKind{Kind: &metadata.MetadataKind_Cluster_{
+					Cluster: &metadata.MetadataKind_Cluster{},
+				}},
+				Name: "istio",
+			},
+			{
+				Kind: &metadata.MetadataKind{Kind: &metadata.MetadataKind_Host_{
+					Host: &metadata.MetadataKind_Host{},
+				}},
+				Name: "istio",
+			},
+		},
+		TransportSocket: &core.TransportSocket{
+			Name:       "envoy.transport_sockets.raw_buffer",
+			ConfigType: &core.TransportSocket_TypedConfig{TypedConfig: protoconv.MessageToAny(&rawbuffer.RawBuffer{})},
+		},
+	})},
+}
+
+func outboundTunnelCluster(proxy *model.Proxy, push *model.PushContext) *cluster.Cluster {
+	return &cluster.Cluster{
+		Name:                 "tunnel",
+		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_ORIGINAL_DST},
+		LbPolicy:             cluster.Cluster_CLUSTER_PROVIDED,
+		ConnectTimeout:       durationpb.New(2 * time.Second),
+		CleanupInterval:      durationpb.New(60 * time.Second),
+		LbConfig: &cluster.Cluster_OriginalDstLbConfig_{
+			OriginalDstLbConfig: &cluster.Cluster_OriginalDstLbConfig{UseHttpHeader: true},
```

Comment:
When will we move off of `x-envoy-original-dst-host` to some standard HBONE header name?

Reply:
The header doesn't actually need to be on the write, this is a hack to pass the original dst through the internal listener. I feel like we already solve that with a custom filter... did we?

Reply:
Yeah the filter is set_dst_address or similar. It looks at the tunnel metadata

Reply:
https://github.com/istio/istio/blob/82cdfb878992835759121c9bd8fa7c0fda452c75/pilot/pkg/networking/util/util.go#L759

If the internal listener has this, you can get rid of the x-envoy header in the listener and the cluster's `OriginalDstLbConfig` can be default.